### PR TITLE
nvgpu softmax cuda version

### PR DESCRIPTION
**Pull request by fengleitian**
Issue:
https://github.com/NervanaSystems/ngraph/issues/2001

Change the cuda version algorithm to the same as our cpu version:

if refer to CUDNN version, it's from:

```
CUDNN_SOFTMAX_FAST
This implementation applies the straightforward softmax operation.
```
to 

```
CUDNN_SOFTMAX_ACCURATE
This implementation scales each point of the softmax input domain by its maximum value to avoid potential floating point overflows in the softmax evaluation.
```

NOTE1: cudnn version only support reduce_axis = 0;
NOTE2: performance issue because https://github.com/NervanaSystems/ngraph/issues/2013

Result:
before:
```
Comparing results for Softmax_11
[INFO] 2018-11-06T07:15:24z test/util/all_close_f.cpp 75        1.02089e-35 is not close to nan
[INFO] 2018-11-06T07:15:24z test/util/all_close_f.cpp 75        5.4829e-33 is not close to nan
[INFO] 2018-11-06T07:15:24z test/util/all_close_f.cpp 75        4.61425e-24 is not close to nan
[INFO] 2018-11-06T07:15:24z test/util/all_close_f.cpp 75        5.80109e-28 is not close to nan
[INFO] 2018-11-06T07:15:24z test/util/all_close_f.cpp 75        0 is not close to nan
```
after:
```
Comparing results for Softmax_11
  14 [INFO] 2018-11-07T01:44:08z test/util/all_close_f.cpp 75    1.02089e-35 is not close to 1.02034e-35
  15 [INFO] 2018-11-07T01:44:08z test/util/all_close_f.cpp 75    4.61425e-24 is not close to 4.61631e-24
  16 [INFO] 2018-11-07T01:44:08z test/util/all_close_f.cpp 75    8.77843e-30 is not close to 8.78233e-30
  17 [INFO] 2018-11-07T01:44:08z test/util/all_close_f.cpp 75    5.02628e-30 is not close to 5.02361e-30
  18 [INFO] 2018-11-07T01:44:08z test/util/all_close_f.cpp 75    5.10065e-33 is not close to 5.10542e-33
```
close to CPU result now.

### CODE REVIEW

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 1]
```diff
@@ -1569,78 +1569,103 @@ size_t runtime::gpu::CUDAEmitter::build_primitive(const op::MaxPool* node)
...
+                                          1,
                                           block_size_x,
                                           1,
-                                          1, // block dim
+                                          1,
                                           0,
-                                          NULL, // shared mem and stream
-                                          arg_list.data(),
-                                          0)); // arguments
+                                          NULL,
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: use nullptr

File: src/ngraph/runtime/gpu/cuda_emitter.cpp [hunk 2]
```diff
@@ -1569,78 +1569,103 @@ size_t runtime::gpu::CUDAEmitter::build_primitive(const op::MaxPool* node)
...
+                                          args_list,
+                                          0));
```
Comment (rkimballn1) [inline src/ngraph/runtime/gpu/cuda_emitter.cpp]: nullptr here too

File: src/ngraph/runtime/gpu/cuda_emitter.cpp (hunk 1 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: done.

File: src/ngraph/runtime/gpu/cuda_emitter.cpp (hunk 2 above)
Comment (fengleitian) [reply src/ngraph/runtime/gpu/cuda_emitter.cpp]: done.

**Comment by fengleitian** [comment]
@csullivan @AyanmoI 
sure. what's failed on cuda reduce?

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 3]
```diff
@@ -329,6 +280,204 @@ void runtime::gpu::CudaKernelBuilder::get_topk(codegen::CodeWriter& writer,
...
+                writer << "int idx" << last_r_idx << " = 0;\n";
+                // unroll last reduction axis
+                uint32_t unroll_num = 8;
+                writer << "for(; idx" << last_r_idx << " + " << unroll_num << " - 1 < reduce_shape"
+                       << last_r_idx << "; idx" << last_r_idx << " += " << unroll_num << ")\n";
+                writer.block_begin();
+                {
+                    for (int k = 0; k < unroll_num; k++)
+                    {
+                        writer << "input_i = expf(in[reduce_idx] - r_max);\n";
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Would be nice to make a load into shared memory for these reused input elements when the reduced sub-tensor is small enough.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 4]
```diff
@@ -329,6 +280,204 @@ void runtime::gpu::CudaKernelBuilder::get_topk(codegen::CodeWriter& writer,
...
+                    }
+                }
+                writer.block_end();
+                writer << "for(; idx" << last_r_idx << " < reduce_shape" << last_r_idx << "; idx"
+                       << last_r_idx << "++)\n";
+                writer.block_begin();
+                {
+                    writer << "input_i = expf(in[reduce_idx] - r_max) / r_sum;\n";
+                    writer << "out[reduce_idx] = input_i;\n";
+                    //writer << "out[reduce_idx] /= r_sum;\n";
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Here too.

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 5]
```diff
@@ -329,6 +280,204 @@ void runtime::gpu::CudaKernelBuilder::get_topk(codegen::CodeWriter& writer,
...
+                uint32_t unroll_num = 8;
+                writer << "for(; idx" << last_r_idx << " + " << unroll_num << " - 1 < reduce_shape"
+                       << last_r_idx << "; idx" << last_r_idx << " += " << unroll_num << ")\n";
+                writer.block_begin();
+                {
+                    for (int k = 0; k < unroll_num; k++)
+                    {
+                        writer << "input_i = expf(in[reduce_idx] - r_max) / r_sum;\n";
+                        writer << "out[reduce_idx] = input_i;\n";
+                        //writer << "out[reduce_idx] /= r_sum;\n";
```
Comment (csullivan) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: Writes should be non-blocking and so loading the pre-exponentiated value would work, but it seems slightly slower in my tests. Remove commented line.

**Comment by fengleitian** [comment]
@csullivan I don't have a measure yet for accuracy. but should be good.

**Pull request merged**